### ISSUE TITLE
SC4-1-1+SC4-1-2-aria-allowed-attribute

### ISSUE THREAD

**Issue by jkodu** [auto-wcag/auto-wcag#199]
---
name: ARIA allowed attribute
description: | 
  This rule checks that `aria-*` attributes are allowed for the element they are on.

success_criterion:
- 4.1.1 # Parsing (A)
- 4.1.2 # Name, Role, Value (A)

test_aspects:
- DOM Tree
- CSS Styling

authors:
- Anne Thyme Nørregaard
---

### Applicability

Any attribute that is placed on an HTML or SVG element and where the attribute name starts with `aria-`.

### Expectation

The `aria-*` attribute is either listed under [Supported States and Properties](https://www.w3.org/TR/wai-aria-1.1/#states_and_properties) in [WAI-ARIA](https://www.w3.org/TR/wai-aria) for the [semantic role](#semantic-role) of the element that the attribute is placed on, or it is found on the list of [Global States and Properties](https://www.w3.org/TR/wai-aria-1.1/#global_states).

**Note:** [Global States and Properties](https://www.w3.org/TR/wai-aria-1.1/#global_states) are supported by all roles and by all base markup elements.

## Assumptions

*There are currently no assumptions*

## Accessibility support

*There are no major accessibility support issues known for this rule.*

## Background

- [Understanding Success Criterion 4.1.1: Parsing](https://www.w3.org/WAI/WCAG21/Understanding/parsing.html)
- [Understanding Success Criterion 4.1.2: Name, Role, Value
](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html)
- [WAI-ARIA 1.1, Supported States and Properties](https://www.w3.org/TR/wai-aria-1.1/#states_and_properties)
- [WAI-ARIA 1.1, Global States and Properties](https://www.w3.org/TR/wai-aria-1.1/#global_states)

#### Passed example 1

`aria-pressed` attribute is allowed for role `button`.

```html
<div role="button" aria-pressed="false"></div>
```

#### Passed example 2

`aria-pressed` attribute is allowed for `role=button` that is the implicit role for `button` element.

```html
<button aria-pressed="false"></button>
```

#### Passed example 3

Global state that is supported by all base markup elements.

```html
<div aria-hidden="true"></div>
```

#### Failed example 1

`aria-pressed` attribute is not allowed for role `checkbox`.

```html
<div role="checkbox" aria-pressed="false"></div>
```

#### Failed example 2

`aria-pressed` attribute is not allowed for `role=checkbox` that is the implicit role for `input` elements of type `checkbox`.

```html
<input type="checkbox" aria-pressed="false">
```

#### Inapplicable example 1

No attribute that starts with `aria-`.

```html
<div role="region"></div>
```

**Comment by annethyme** [comment]
https://github.com/auto-wcag/auto-wcag/pull/257
